File: app/src/main/java/me/zhanghai/android/files/provider/remote/ParcelablePathListConsumer.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.remote;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import java8.nio.file.Path;
import java9.util.function.Consumer;
import me.zhanghai.android.files.util.BundleBuilder;
import me.zhanghai.android.files.util.BundleUtils;
import me.zhanghai.android.files.util.CollectionUtils;
import me.zhanghai.android.files.util.RemoteCallback;

public class ParcelablePathListConsumer implements Parcelable {

    private static final String KEY_PREFIX = ParcelablePathListConsumer.class.getName() + '.';

    private static final String KEY_PATH_LIST = KEY_PREFIX + "PATH_LIST";

    @NonNull
    private final Consumer<List<Path>> mListener;

    public ParcelablePathListConsumer(@NonNull Consumer<List<Path>> listener) {
        mListener = listener;
    }

    @NonNull
    public Consumer<List<Path>> get() {
        return mListener;
    }


    public static final Creator<ParcelablePathListConsumer> CREATOR =
            new Creator<ParcelablePathListConsumer>() {
                @Override
                public ParcelablePathListConsumer createFromParcel(Parcel source) {
                    return new ParcelablePathListConsumer(source);
                }
                @Override
                public ParcelablePathListConsumer[] newArray(int size) {
                    return new ParcelablePathListConsumer[size];
                }
            };

    protected ParcelablePathListConsumer(Parcel in) {
        RemoteCallback remoteCallback = in.readParcelable(RemoteCallback.class.getClassLoader());
        //noinspection unchecked
        mListener = paths -> remoteCallback.sendResult(new BundleBuilder()
                .putParcelableArrayList(KEY_PATH_LIST,
                        (ArrayList<Parcelable>) (ArrayList<?>) CollectionUtils.toArrayList(paths))
                .build());
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        //noinspection unchecked
        dest.writeParcelable(new RemoteCallback(result -> mListener.accept(
                (List<Path>) (List<?>) BundleUtils.getParcelableArrayList(result, KEY_PATH_LIST))),
                flags);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/remote/ParcelablePathListConsumer.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.remote

import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import java8.nio.file.Path
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.WriteWith
import me.zhanghai.android.files.util.ParcelableArgs
import me.zhanghai.android.files.util.ParcelableListParceler
import me.zhanghai.android.files.util.RemoteCallback
import me.zhanghai.android.files.util.getArgs
import me.zhanghai.android.files.util.putArgs

class ParcelablePathListConsumer(val value: (List<Path>) -> Unit) : Parcelable {
    private constructor(source: Parcel, loader: ClassLoader?) : this(
        source.readParcelable<RemoteCallback>(loader)!!.let<RemoteCallback, (List<Path>) -> Unit> {
            { paths -> it.sendResult(Bundle().putArgs(ListenerArgs(paths))) }
        }
    )

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(RemoteCallback { value(it.getArgs<ListenerArgs>().paths) }, flags)
    }

    companion object {
        @JvmField
        val CREATOR = object : Parcelable.ClassLoaderCreator<ParcelablePathListConsumer> {
            override fun createFromParcel(source: Parcel): ParcelablePathListConsumer =
                createFromParcel(source, null)

            override fun createFromParcel(
                source: Parcel,
                loader: ClassLoader?
            ): ParcelablePathListConsumer = ParcelablePathListConsumer(source, loader)

            override fun newArray(size: Int): Array<ParcelablePathListConsumer?> =
                arrayOfNulls(size)
        }
    }

    @Parcelize
    private class ListenerArgs(
        val paths: @WriteWith<ParcelableListParceler> List<Path>
    ) : ParcelableArgs
}

fun ((List<Path>) -> Unit).toParcelable(): ParcelablePathListConsumer =
    ParcelablePathListConsumer(this)
